ccp-openshift-scan/scanning/misc_package_updates.py

```diff
@@ -0,0 +1,140 @@
+    """
+    command = [executable, "list", "--outdated", "--disable-pip-version-check"]
+    out, err = [], ""
+
+    try:
+        out, err = lib.run_cmd_out_err(command)
+    except Exception as e:
+        err = e
+
+    if err:
+        if binary_does_not_exist(err):
+            return "{} is not installed".format(executable)
+        else:
+            return "Failed to find the pip updates."
+    else:
+        if out.strip():
+            return out.strip().split("\n")
+        else:
+            return []
+
+
+def find_npm_updates(executable="/usr/bin/npm"):
+    """
+    Finds out outdated installed packages of npm
+    """
+    command = [executable, "-g", "outdated"]
+    out, err = [], ""
+
+    try:
+        out, err = lib.run_cmd_out_err(command)
+    except Exception as e:
+        err = e
+
+    if err:
+        if binary_does_not_exist(err):
+            return "{} is not installed".format(executable)
+        else:
+            return "Failed to find the npm updates."
+    else:
+        if out.strip():
+            return out.strip().split("\n")
+        else:
+            return []
+
+
+def find_gem_updates(executable="/usr/bin/gem"):
+    """
+    Finds out outdated installed packages of gem
+    """
+    command = [executable, "outdated"]
+    out, err = [], ""
+
+    try:
+        out, err = lib.run_cmd_out_err(command)
+    except Exception as e:
+        err = e
+
+    if err:
+        if binary_does_not_exist(err):
+            return "{} is not installed".format(executable)
+        else:
+            return "Failed to find the gem updates."
+    else:
+        if out.strip():
+            return out.strip().split("\n")
+        else:
+            return []
+
+
+def print_updates(binary):
+    """
+    Prints the updates found using givn binary
+    """
+    print ("\n{} updates scan:".format(binary))
+
+    if binary == "npm":
+        result = find_npm_updates()
+    elif binary == "gem":
+        result = find_gem_updates()
+    elif binary == "pip":
+        result = find_pip_updates()
+    else:
+        return
+
+    if result:
+        # prints errors
+        if isinstance(result, str):
+            print (result)
+            return
+        # prints result
+        for line in result:
+            print (line)
+    else:
+        print ("No updates required.")
+
+
+if __name__ == "__main__":
+    valid_args = ["pip", "gem", "npm", "all"]
+
+    if len(sys.argv) < 2:
+        print ("Please provide at least one argument as:")
+        print ("python misc_package_updates.py npm")
+        print ("Valid arguments: {}".format(valid_args))
+        sys.exit(1)
+
+    cli_arg = sys.argv[1].strip()
+
+    if cli_arg not in valid_args:
+        print ("Please provide valid args among {}".format(valid_args))
+        sys.exit(1)
+
+    if cli_arg == "all":
+        print_updates("pip")
+        print_updates("npm")
+        print_updates("gem")
+    else:
+        print_updates(cli_arg)
```

Comment:
We have 3 defined package manager choices. The `print_updates` takes care of printing updates for one package manager. We just re-use it 3 times.
Having provision to take `all` option in `print_updates` would require it to loop over the defined number of package managers, so I chose not to complicate the print_updates function for defined number of iterations.

Reply:
We don't really need to loop over them, right? We can call the three functions `find_{pip,npm,gem}_updates` inside `print_updates` function. 

I don't know too much about the internals, but in current implementation, we will have 6 function calls when we need results for all package managers. But if we make `print_updates` understand `all`, it will be 4 function calls.

Reply:
@navidshaikh Can you reply?

Reply:
I chose to not add complexity to `print_updates` method and use `all` method as the wrapper method which individually calls others package manager updates utilities.